### PR TITLE
Fix jvm heap memory leak of JDBC scanner

### PR DESCRIPTION
**Pull request by miomiocat**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
Fixes #

## Problem Summary(Required) ：

jni local reference need to be deleted after use, otherwise it will cause heap memory leak and cannot be GCed.

GC throughput can be improved from 40% to 97% in our cases.

### CODE REVIEW

**Review by DorianZheng — DISMISSED** [review]
generally LGTM

**Review by imay — DISMISSED** [review]

File: be/src/exec/vectorized/jdbc_scanner.cpp [hunk 1]
```diff
@@ -439,21 +442,29 @@ Status JDBCScanner::_fill_chunk(jobject jchunk, ChunkPtr* chunk) {
         } else if (column_class == "java.sql.Timestamp") {
             DCHECK(slot_desc->type().type == TYPE_DATETIME);
             for (int i = 0; i < num_rows; i++) {
-                RETURN_IF_ERROR(_append_datetime_val(helper.list_get(jcolumn, i), slot_desc, column.get()));
+                jobject jelement = helper.list_get(jcolumn, i);
+                DeferOp defer([&jelement, this]() { _jni_env->DeleteLocalRef(jelement); });
```
Comment (stdpain) [inline be/src/exec/vectorized/jdbc_scanner.cpp:446]: it seems jelement will be delete in _append_datetime_val 
PTAL @silverbullet233

File: be/src/exec/vectorized/jdbc_scanner.cpp [hunk 2]
```diff
@@ -414,13 +416,14 @@ Status JDBCScanner::_fill_chunk(jobject jchunk, ChunkPtr* chunk) {
...
-            RETURN_IF_ERROR(                                                                                          \
-                    _append_value_from_result<cpp_type>(helper.list_get(jcolumn, i), func, slot_desc, column.get())); \
-        }                                                                                                             \
+#define FILL_COLUMN(get_value_func, cpp_type)                                                              \
+    {                                                                                                      \
+        auto func = std::bind(&JVMFunctionHelper::get_value_func, &helper, std::placeholders::_1);         \
+        for (int i = 0; i < num_rows; i++) {                                                               \
+            jobject jelement = helper.list_get(jcolumn, i);                                                \
+            DeferOp defer([&jelement, this]() { _jni_env->DeleteLocalRef(jelement); });                    \
+            RETURN_IF_ERROR(_append_value_from_result<cpp_type>(jelement, func, slot_desc, column.get())); \
```
Comment (silverbullet233) [inline be/src/exec/vectorized/jdbc_scanner.cpp:425]: I think it's better to put this DeferOp inside _append_value_from_result, just like other _append_*_val functions

File: be/src/exec/vectorized/jdbc_scanner.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/vectorized/jdbc_scanner.cpp:446]: > it seems jelement will be delete in _append_datetime_val PTAL @silverbullet233

yes, I think this defer is no need

**Comment by silverbullet233** [comment]
There is already a DeferOp in each `_append_*_val` function used to delete local ref, I think it's not necessary to do it again externally. @miomiocat

**Comment by miomiocat** [comment]
> There is already a DeferOp in each `_append_*_val` function used to delete local ref, I think it's not necessary to do it again externally. @miomiocat

DeferOp is missing for the jobect used in `_append_value_from_result`, PTAL

File: be/src/exec/vectorized/jdbc_scanner.cpp (hunk 2 above)
Comment (miomiocat) [reply be/src/exec/vectorized/jdbc_scanner.cpp]: done

**Comment by silverbullet233** [comment]
> > There is already a DeferOp in each `_append_*_val` function used to delete local ref, I think it's not necessary to do it again externally. @miomiocat
> 
> DeferOp is missing for the jobect used in `_append_value_from_result`, PTAL

LGTM

**Review by silverbullet233 — APPROVED** [review]

**Comment by miomiocat** [comment]
run starrocks_be_unittest

**Review by dirtysalt — APPROVED** [review]

**Comment by stdpain** [comment]
you can pass -Xcheck:jni to check refs leak

**Review by stdpain — APPROVED** [review]

**Pull request merged**